src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+   */
+  private nextVersion: Map<VersionId, VersionId> = new Map();
+
+  /**
+   * Map a operation hash to the OperationInfo which contains sufficient
+   * information to reconstruct the operation.
+   */
+  private opHashToInfo: Map<OperationHash, OperationInfo> = new Map();
+
+  /**
+   * Apply (perform) a specified DID state changing operation.
+   */
+  public apply (operation: WriteOperation): string | null {
+    const opHash = DidCache.getHash(operation);
+
+    // Ignore operations without the required metadata - any operation anchored
+    // in a blockchain should have this metadata.
+    if (operation.transactionNumber === undefined ||
+        operation.operationIndex === undefined ||
+        operation.batchFileHash === undefined) {
+      return null;
+    }
+
+    // opInfo is operation with derivable properties projected out
+    const opInfo: OperationInfo = {
+      transactionNumber: operation.transactionNumber,
+      operationIndex: operation.operationIndex,
+      batchFileHash: operation.batchFileHash,
+      type: operation.type
+    };
+
+    // If this is a duplicate of an earlier operation, we can
+    // ignore this operation. Note that we might have a previous
+    // operation with the same hash, but that previous operation
+    // need not be earlier in timestamp order - hence the check
+    // with lesser().
+    const prevOperation = this.opHashToInfo.get(opHash);
+    if (prevOperation !== undefined && lesser(prevOperation, opInfo)) {
+      return null;
+    }
+    // Update our mapping of operation hash to operation info overwriting
+    // previous info if it exists
+    this.opHashToInfo.set(opHash, opInfo);
+
+    // For operations that have a previous version, we need additional
+    // bookkeeping
+    if (operation.previousOperationHash) {
+      this.applyOpWithPrev(opHash, opInfo, operation.previousOperationHash);
+    }
+
+    return opHash;
+  }
+
+  public constructor (
+    private readonly cas: Cas,
+    private readonly didDocUpdate: DidDocumentUpdater,
+    private readonly didDocCreate: DidDocumentCreator) {
+
+  }
+
+  /**
+   * Rollback the state of the DidCache by removing all operations
+   * with transactionNumber greater than the provided parameter value.
+   * The intended use case for this method is to handle rollbacks
+   * in the blockchain.
+   *
+   * The current implementation is inefficient: It simply scans the two
+   * hashmaps storing the core Did state and removes all entries with
+   * a greater transaction number.  In future, the implementation should be optimized
+   * for the common case by keeping a sliding window of recent operations.
+   */
+  public rollback (transactionNumber: number) {
+
+    // Iterate over all nextVersion entries and remove all versions
+    // with "next" operation with transactionNumber greater than the provided
+    // parameter.
+    this.nextVersion.forEach((opHash, version, map) => {
```

Comment:
(version, nextVersion, map) to be consistent, since the code makes a point of making a distinction.

Reply:
Fixed per suggestion.